docs/dev/relay_add_pass.rst

```diff
@@ -0,0 +1,341 @@
+        return tuple->fields[op->index];
+      } else {
+        return res;
+      }
+    }
+
+In the ``TupleItemGetNode`` case, we check if ``op->tuple`` field is a
+``TupleNode``. If so, we replace the tuple get with the field of the tuple
+pointed to by ``op->index``. The reason we need to check is because
+``op->tuple`` might evaluate to a tuple, without itself being a tuple.
+
+.. code:: c
+
+    Expr VisitExpr_(const CallNode* call) final {
+      static auto op_stateful = Op::GetAttr<TOpIsStateful>("TOpIsStateful");
+      Expr res = ExprMutator::VisitExpr_(call);
+      call = res.as<CallNode>();
+      // We don't constant fold function with zero arguments.
+      // This is a heuristic that is useful.
+      // For example it is harmful to fold ones(shape=(4, 5)).
+      if (call->args.size() == 0) return res;
+      const OpNode* op = call->op.as<OpNode>();
+      if (op == nullptr) return res;
+      // skip stateful ops.
+      if (op_stateful.get(GetRef<Op>(op), false)) return res;
+      bool all_const_args = true;
+      for (Expr arg : call->args) {
+        if (!checker_.Check(arg)) {
+          all_const_args = false;
+        }
+      }
+      if (all_const_args) {
+        return ConstEvaluate(res);
+      } else {
+        return res;
+      }
+    }
+
+In the ``CallNode`` case, we first use the ``VisitExpr_`` of ``ExprMutator``
+to visit the call, which const-folds all of the fields of the call. We use
+``ExprMutator::VisitExpr_`` instead of ``VisitExpr``, because we want to
+bypass the vtable (to avoid an infinite loop) and use the default
+implementation provided by ``ExprMutator``. Then we evaluate the call only if
+all of the arguments are constant (using ``ConstantChecker``). Evaluating the
+call produces a **value**, so we use a helper method ``ValueToExpr`` to allow
+us to place the evaluated expression back into the AST.
+
+Now, we construct the public interface ``FoldConstant`` to our constant
+folder, which is a standalone function outside of the ``ConstantFolder``
+class. ``FoldConstant`` takes an expression and internally creates and uses a
+``ConstantFolder`` instance (the full definition can be found in
+``include/tvm/relay/pass.h``).
+
+To allow other C++ modules to use our pass, we declare the public interface
+in ``src/relay/pass/pass.h``:
+
+.. code:: c
+
+    TVM_DLL Expr FoldConstant(const Expr& expr);
+
+Registering an API Endpoint
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the AST traversers written, the pass can be registered to become a TVM
+API endpoint with the following code snippet:
+
+.. code:: c
+
+    TVM_REGISTER_API("relay._ir_pass.FoldConstant")
+    .set_body([](TVMArgs args, TVMRetValue *ret) {
+        *ret = FoldConstant(args[0]);
+    });
+
+And the pass can now be used in C++ and Python, though it's a good idea to
+wrap the API in Python, as described in :ref:`relay-add-op`. More detail
+about registration can be found in :doc:`runtime.md`.
```

Comment:
One problem when building these docs is that the `:doc:runtime.md` reference doesn't get resolved.  And we can't use the normal `:ref:ayy-lmao` mechanism to point to those docs, because they're in Markdown, and not RST.

Reply:
Can we just rewrite it into rst?

Reply:
I've found it successful to add explicit references into docs and use `:ref:` to point to those references (see the Relay lang ref for examples)

Reply:
> Can we just rewrite it into rst?

Yeah.  That seems like the way to go.  Gonna rewrite it in ~rust~ rst today.

Reply:
So uhh... why are the docs written in rst?  It doesn't seem to highlight C or C++ very well and figuring out how to use its syntax feels like a puzzle.